### PR TITLE
Task/NATs Message Queue Trigger

### PR DESCRIPTION
**Pull request by conbrad**
# Test Links:
[Percentile Calculator](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Fire Behaviour Advisory](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-2261.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

**Comment by conbrad** [comment]
Closing to save resources

File: api/app/auto_spatial_advisory/sfms.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    now = get_vancouver_now()
+    now_date = now.date()
+
+    if file_date < now_date:
+        # It's from the past - it's an actual.
+        return True
+    if file_date > now_date:
+        # It's from the future - it's a forecast.
+        return False
+    # It's from today - now it get's weird.
```
Comment (andrea-williams) [inline api/app/auto_spatial_advisory/sfms.py:37]:
```suggestion
    # It's from today - now it gets weird.
```

Comment (andrea-williams) [reply api/app/auto_spatial_advisory/sfms.py:37]: what does my 20 points get me???

Comment (Sybrand) [reply api/app/auto_spatial_advisory/sfms.py:37]: Once you have 100 points, you can trade it for a virtual gold piece!

File: api/app/nats.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+#     loop.run_until_complete(test_basic_messaging_functions())
+
+
+# async def test_basic_messaging_functions():
+#     nc = await nats.connect("localhost")
+
+#     # Create JetStream context.
+#     js = nc.jetstream()
+
+#     # Persist messages on 'foo's subject.
```
Comment (andrea-williams) [inline api/app/nats.py:73]: what's the plan with all this commented-out code?

Comment (Sybrand) [reply api/app/nats.py:73]: Figured it's a good reference for the next ticket - which is consuming the messages. I could delete it now - but for sure #2224 should delete it.

File: api/app/auto_spatial_advisory/sfms.py (hunk 1 above)
Comment (andrea-williams) [reply api/app/auto_spatial_advisory/sfms.py]: a bitcoin?!?!?

Comment (Sybrand) [reply api/app/auto_spatial_advisory/sfms.py]: Ok fine. I'll get you a bitcoin when it reaches parity with the Zimbabwean dollar.

File: api/app/auto_spatial_advisory/sfms.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def get_target_filename(filename: str) -> str:
+    """ Get the target filename, something that looks like this:
+    bucket/sfms/upload/forecast/[issue date NOT TIME]/hfi20220823.tif
+    bucket/sfms/upload/actual/[issue date NOT TIME]/hfi20220823.tif
+    """
+    # We are assuming that the local server time, matches the issue date. We assume that
+    # right after a file is generated, this API is called - and as such the current
+    # time IS the issue date.
```
[the remark this answers is not in the document]
Comment (Sybrand) [reply api/app/auto_spatial_advisory/sfms.py:57]: I have agonized over this very matter. It's supposed to happen right away. There is the Last-modified and Create-time, which may be a better choice.
I don't know what time zone the server is configured to be - but those may be a better thing to go with than the current server time.

File: api/app/routers/sfms.py [hunk 4]
```diff
@@ -120,16 +74,34 @@ async def upload(file: UploadFile,
     """
     logger.info('sfms/upload/')
     secret = request.headers.get('Secret')
-    if secret != config.get('SFMS_SECRET'):
+    if secret == '' or secret != config.get('SFMS_SECRET'):
```
[the remark this answers is not in the document]
Comment (Sybrand) [reply api/app/routers/sfms.py:77]: Aaah - let me change that to:
```suggestion
    if secret or secret != config.get('SFMS_SECRET'):
```
What I really want to catch is if you didn't specify a secret, or if the secret doesn't match. (But I never want None or '' to be valid secrets either)

**Review by andrea-williams — APPROVED** [review]

File: api/app/auto_spatial_advisory/sfms.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def get_date_part(filename: str) -> str:
+    """ Get the date part of the filename.
+    Filename example: hfi20220823.tif
+    """
+    return filename[filename.rfind('.') - 8:filename.rfind('.')]
+
+
+def is_actual(filename: str) -> bool:
```
Comment (conbrad) [inline api/app/auto_spatial_advisory/sfms.py:16]: Minor: could simplify this to return `actual` or `forecast` and kill the `get_prefix` function to simplify

File: api/app/tests/sfms/test_sfms_upload.py [hunk 6]
```diff
@@ -46,7 +55,13 @@ def test_is_actual_before_noon(_):
     assert is_actual('hfi20220824.tif') is False
 
 
-@patch('app.routers.sfms.get_vancouver_now', return_value=get_pdt_noon())
+def test_tiff_does_not_break():
+    """ If someone decided to add an extra f on tiff, it should keep working!"""
```
Comment (conbrad) [inline api/app/tests/sfms/test_sfms_upload.py:59]: 💯

**Review by Sybrand — APPROVED** [review]

**Pull request merged**